compiler/rustc_mir/src/monomorphize/collector.rs

```diff
@@ -491,12 +491,19 @@ fn check_type_length_limit<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) {
             // This also avoids the case where before and after slices overlap.
             if shrunk.len() < s.len() { shrunk } else { s }
         };
+        let instance_str = instance.to_string();
+        let temp_path = tcx.output_filenames(LOCAL_CRATE).temp_path_ext("long-type.txt", None);
+        let written_ok = std::fs::write(&temp_path, &instance_str).is_ok();
```

Comment:
We should only write the file if the shrunken string is shorter than the full version above. I think the`$CLOSURE` case below hits this.

Reply:
The test doesn't hit this (the type is still slightly longer), but I've added the logic to skip truncation and file write on short types.